CL-704 Representativeness score (computed on the fly)

adessy opened the pull request:
See CL-704 for the description.

I'm hesitating to rename:
- `.../users/custom_field/:custom_field_id/representativeness` to `.../users/custom_field/:custom_field_id/rscore` and
- the `type` in responses from `representativeness_score` to `rscore`

Happy reviewing :octocat:

adessy committed:
CL-704 Extract logic to find/filter users into a new Finder object

This logic will be reused to select groups of users for the computation of representativeness scores.

adessy committed:
CL-704 Extract logic to count the nb of users by custom field option

This logic will be reused to compute representativeness scores.

adessy committed:
CL-704 Add documentation to UserCustomFields::FieldValueCounter

adessy committed:
CL-704 Make the factory for reference distribution deterministic

It also makes it more convenient to tune the population counts.

adessy committed:
CL-704 Implement the representativeness score pseudo-model

For now, representativeness scores are not persisted in the database. That's the reason why RScore is not implemented as an ActiveRecord model, but this will be changed in the next iteration.

adessy committed:
CL-704 Use `RSpec.describe` instead of `describe`

adessy committed:
CL-704 Store user_counts in RScore's

RScore instances stores parameters that were used to compute them, namely user_counts and ref_distribution. This will be used for debug and audit purposes. This will also allow us to recompute the scores with a different scoring algorithm in the future if necessary.

adessy committed:
CL-704 Add representativeness score endpoint

adessy committed:
CL-704 Add some tests for UserCustomFields::Representativeness::RScorePolicy

adessy committed:
CL-704 Fix rubocop offenses

adessy committed:
CL-704 Log representativness score values for alternative scoring methods

adessy committed:
CL-704 Merge branch 'master' into CL-704/representativeness-score-resource

adessy committed:
CL-704 Fix the reference-distribution factory

If the custom field had more than 2 options, the distribution would include nil counts. This is now fixed by cycling through the population counts if necessary. Ideally, we would check that the population_counts size is equal to the number of options when both the custom field and the population_counts are provided explicitly, but this is not straightforward to implement with FactoryBot API.

I also added some extra validation rules to make it easier to identify invalid distributions in the future.

adessy committed:
CL-704 Include the reference distribution in rscore responses

adessy committed:
CL-704 Test that reference distribution are included in RScore responses

adessy committed:
CL-704 Add support for `project` filter to representativeness endpoint

adessy committed:
CL-704 Rename `CustomField#custom_field_options` to `CustomField#options`

luucvanderzee commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/r_score.rb:
```diff
@@ -0,0 +1,101 @@
+      population_counts = ref_distribution.distribution
+      check_user_counts_options!(user_counts, ref_distribution)
+
+      score_value = compute_scores(user_counts, population_counts)[:min_max_p_ratio]
```
Do you want to use the min/max score in the dashboard? I thought we would do the min/mean score because it's a bit less pessimistic

adessy replied:
I have no strong preference. The min/max is a bit more standard, so I thought I would give it a chance first. But really, whatever.

jamesspeake replied:
If min/max is more standard I think I agree we should give it a chance - then look at what the real data looks like.

luucvanderzee commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/r_score.rb:
```diff
@@ -0,0 +1,101 @@
+    def check_user_counts_options!(user_counts, ref_distribution)
+      known_options = ref_distribution.distribution.keys + [UserCustomFields::FieldValueCounter::UNKNOWN_VALUE_LABEL]
+      unknown_options = user_counts.keys - known_options
+      return if unknown_options.blank?
```
So if I understand correctly, if the union of `ref_distribution.distribution.keys` and `["_blank"]` has less keys than `user_counts`, we throw an error. That does not seem desirable right? The `ref_distribution.distribution` (at least how I have implemented it now on the FE) will often have keys that are not in `user_counts`- if the admin wants to disable certain options they will not be in the reference distribution object and this would throw an error. Or am I missing something?

I used the following dummy code to test this scenario:

```
reference_dist = { "a" => 1, "b" => 2 }
known_options = reference_dist.keys + ["_blank"]

user_counts = { "a" => 10, "b" => 20, "c" => 30, "_blank" => 25 }

unknown_options = user_counts.keys - known_options

if (unknown_options.length != 0)
   puts "ERROR" 
end
```

In this case, the reference distribution is missing the `c` key (which would mean that the admin disabled that option), so this is a valid use case and should not throw an error, but I still get `"ERROR"` logged to the console

adessy replied:
You're totally right. I forgot about the toggle thing here.

luucvanderzee commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/ref_distribution.rb:
```diff
@@ -44,6 +45,17 @@ def validate_distribution_options
+    errors.add(:distribution, 'population counts cannot be nil.') if counts.any?(&:nil?)
+
+    counts.compact!
+    errors.add(:distribution, 'population counts must be strictly positive.') unless counts.all?(&:positive?)
```
Ok, good to know that values cannot be zero- forgot about this case, will also update this on the FE

adessy replied:
Yeah I thought having both 0 and the toggle would be confusing for everyone, including me 😅

luucvanderzee commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/r_score.rb:
```diff
@@ -0,0 +1,101 @@
+    end
+
+    def participation_rates(user_counts, population_counts)
+      population_counts.keys.map do |option_id|
```
Related to my comment below regarding `check_user_counts_options!`:

If we use `population_counts.keys` to `.map` over, and I assume that the user distribution can have keys that are not in the reference distribution (because of the admin choosing to disable fields, which would currently throw an error in `check_user_counts_options!`), we can get the following situation: 

All the fields that the admin has selected happen to have a user count of zero. In this case, all `participation_rates` are zero. If you then calculate the `min_max_p_ratio` or the `min_mean_p_ratio` you are dividing by zero. Maybe there should be an extra check to make sure this is not the case? I was thinking of adding this check in the FE as well just for the population data because if the absolute population values are all 0 the percentage calculation would get messed up as well

luucvanderzee replied:
By the way, might be more related to the previous PR, but how are you calculating the `expected_users` exactly? Because I think we should only calculate the expected users based on the keys that are actually in the reference distribution, which I think is not the case right now. In the question I asked you yesterday about the `expected_users` (see image) the expected user counts are ` { male: 3.8, female: 4.2 }` even though the absolute numbers of users in the dashboard are ` { male: 1, female: 1 }` (because the reference distribution has no value for `unspecified`). Shouldn't the sum of expected user counts be equal to the total user count, i.e. ignore `unspecified` when it's calculated?

Sorry really choatic story haha, hope it makes sense!

![image](https://user-images.githubusercontent.com/13822005/175038135-1577eda1-e222-4558-9759-5e9a2b93a402.png)

adessy replied:
> If you then calculate the min_max_p_ratio or the min_mean_p_ratio you are dividing by zero.

Good catch! that toggle...

About your second point, it's an interesting question. I guess it depends on the semantic of the toggle. If the toggle is not assimilated to a zero, it probably makes indeed more sense to not include the counts of disabled options, especially if they are not shown in UI.

jamesspeake commented on back/engines/commercial/user_custom_fields/app/policies/user_custom_fields/representativeness/r_score_policy.rb:
```diff
@@ -0,0 +1,11 @@
+  module Representativeness
+    class RScorePolicy < ApplicationPolicy
+      def show?
+        active? && admin?
```
Is this definitely only available to admins?

adessy replied:
At least for now :) It might make sense to add moderators as well, but then we need to define what should be shown by default when opening the dashboard and no project has been selected yet.

jamesspeake commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/r_score.rb:
```diff
@@ -0,0 +1,101 @@
+        proportional_similarity: proportional_similarity(user_counts, population_counts)
+      }
+
+      Rails.logger.info(
```
Is this logging in a way that can be easily extracted for analysis? Do 'info' statements get logged on production?

adessy replied:
It's the first time I'm doing this, so it's kind of an experiment and I was planning to do an early check after the release. My main concern is how long the logs are kept.

> Do 'info' statements get logged on production?

Yes!

jamesspeake reviewed (commented):
I've been through all the files and think I've understood most of what it is doing. Beyond the couple of comments and agreeing with Luuc on the couple of issues that need addressing, I think we need to get this somewhere so we can start looking at what it actually produces.

adessy committed:
CL-704 Ensure that scoring methods returns NaN when ill-defined

The representativeness score is considered ill-defined when there is no participants for any option described in the reference distribution. In that case, the method must return NaN.

adessy committed:
CL-704 Merge branch 'master' into CL-704/representativeness-score-resource

# Conflicts:
#	back/engines/commercial/user_custom_fields/app/controllers/user_custom_fields/web_api/v1/stats_users_controller.rb

adessy committed:
CL-704 Include options with 0 counts in the `users` series of user stats endpoints

adessy committed:
CL-704 Rename `CustomField#custom_field_options` to `CustomField#options`

Merge from master reintroduced some occurrences of `CustomField#custom_field_options`.

adessy committed:
CL-704 Remove buggy consistency check

I removed `check_user_counts_options!` which was too strict. That's because in the current implementation keys of options that have been toggled off (when creating the reference distribution) do not appear in the hash of reference distribution.

adessy committed:
CL-704 Fix buggy consistency check when adding missing options

The consistency check could only work for custom fields with reified (not implicit) options. Skipping custom fields without explicit options for now as it was more of a nice to have.

adessy committed:
CL-704 Rework the calculation of expected user counts

Only user counts for options toggled on in the reference distribution are used to compute the expected user counts.

adessy commented:
I reworked a few things:
- Only user counts for options toggled on in the reference distribution are used to compute the expected user counts.
- Scoring methods return NaN for ill-defined scores (0 participants). 
- I removed some consistency checks that were too strict because of the toggle-off feature as pointed out in review comments.
- `user` series returned by `.../users_by...` endpoints now include zero counts for options that are not represented in the selected subset of users (only works for custom fields with options – does not cover number, checkbox, domicile).

Last question: do you prefer `.../users/custom_field/:custom_field_id/representativeness` or `.../users/custom_field/:custom_field_id/rscore`?

luucvanderzee commented:
> I removed some consistency checks that were too strict because of the toggle-off feature as pointed out in review comments.

So is the expected structure of the reference distribution payload still the same? I.e. only fields that are enabled and have a key-value pair in the object, and disabled fields don't have a key-value pair in the object?

> user series returned by .../users_by... endpoints now include zero counts for options that are not represented in the selected subset of users (only works for custom fields with options – does not cover number, checkbox, domicile).

I wonder if it's not easier to just keep things the way they were, so no zero counts in the response. If this works differently for different fields that only adds complexity on the front-end. Maybe we should only include the zero counts when we do it for all fields?

> Last question: do you prefer .../users/custom_field/:custom_field_id/representativeness or .../users/custom_field/:custom_field_id/rscore?

I like `rscore` better!

luucvanderzee commented on back/engines/commercial/user_custom_fields/app/models/user_custom_fields/representativeness/r_score.rb:
```diff
@@ -0,0 +1,90 @@
+    # @return [UserCustomFields::Representativeness::RScore]
+    def compute(user_counts, ref_distribution)
+      population_counts = ref_distribution.distribution
+      score_value = compute_scores(user_counts, population_counts)[:min_max_p_ratio]
```
I thought we were going for the `min_mean_p_ratio`?

adessy replied:
Have you seen this thread: https://github.com/CitizenLabDotCo/citizenlab/pull/2182#discussion_r903686457?

jamesspeake replied:
Yes think stick to min/max for now

adessy commented:
> So is the expected structure of the reference distribution payload still the same? I.e. only fields that are enabled and have a key-value pair in the object, and disabled fields don't have a key-value pair in the object?

For now yes, we didn't want to delay this any further. It might be something we reconsider in the future.

> I wonder if it's not easier to just keep things the way they were, so no zero counts in the response. If this works differently for different fields that only adds complexity on the front-end. Maybe we should only include the zero counts when we do it for all fields?

Is it really changing anything on the front-end side? My guess was that the current implementation considers the attribute optional and interprets it as 0 when missing. CL-959 should fix it for domicile. I was planning to implement it for checkboxes later, but number field is still a bit of a question mark. It probably doesn't make sense to list all the missing options for the latter.

> I like rscore better!

👍

jamesspeake commented:
On the point of zero counts in the response, @luucvanderzee does it add a lot of effort to have different implementations if we are moving to using it for all fields?
 
I like rscore better too

jamesspeake reviewed (approved):
I'm happy with the changes pending the comment about zero values that Luuc has raised. Given that is the direction of travel, I wouldn't change unless a significant overhead on the front end.

adessy committed:
CL-704 Rename representativeness score endpoint

adessy committed:
CL-704 Change the representativeness score type to `rscore`

luucvanderzee commented:
The way I have implemented it now it just checks if all the fields that are in the `reference_distribution` are also in the response. If the field is there it's used, and if it's not it adds it with `0`. So that covers all cases I think :)
